### ISSUE TITLE
`unnecessary_sort_by` false positive or wrong suggestion

### ISSUE THREAD

**Issue by awused** [rust-lang/rust-clippy#7822]
Lint name: unnecessary_sort_by

I tried this code: 

```rust
    let mut a: Vec<Ref<usize>> = Vec::new();

    a.sort_unstable_by(|a, b| a.cmp(b));
```

I expected to see this happen: Nothing

Instead, this happened: clippy suggests

```rust
    let mut a: Vec<Ref<usize>> = Vec::new();

    a.sort_unstable();
```

This does not build because Ord is not implemented for Ref<'_, T: Ord>.

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.55.0 (c8dfcfe04 2021-09-06)
binary: rustc
commit-hash: c8dfcfe046a7680554bf4eb612bad840e7631c4b
commit-date: 2021-09-06
host: x86_64-unknown-linux-gnu
release: 1.55.0
LLVM version: 12.0.1
```

@rustbot label +I-suggestion-causes-error
